plugins/nextcloud-rule-exclusions-before.conf

```diff
@@ -463,8 +464,19 @@ SecRule REQUEST_FILENAME "@contains /index.php/apps/notes/" \
     t:none,\
     nolog,\
     ctl:ruleRemoveByTag=attack-injection-php,\
-    ver:'nextcloud-rule-exclusions-plugin/1.0.0'"
+    ver:'nextcloud-rule-exclusions-plugin/1.0.0',\
+    setvar:'tx.allowed_methods=%{tx.allowed_methods} PUT'"
 
+# Allow Notes App to save to its settings
+SecRule REQUEST_FILENAME "@contains /apps/notes/settings" \
```

Comment:
Is this path really distinct from the on in `9508340`?

Reply:
Apparently. Is it possible some proxy configuration is responsible for changing the URI and is it possible (and desireable) to catch `@contains /apps/notes/settings` in any path?

Reply:
Its possible but unlikely that a proxy configuration manipulated the path like that. If the two rules can be combined by using `@contains /apps/notes/settings` then I would do that. Every additional rule adds a tiny overhead to request processing and duplication is bad for maintenance. But that really depends on whether there are other contexts that might be affected by the rule. Alternatively, you could change the rule to say `@rx (?:/index.php)?/apps/notes(?:/settings)?` to make it fit your requirements better (*Do not* use the expression I wrote, it's just an example. It would also match `/apps/notes` and you might not want that).

Reply:
There is a rule for the notes app:

    # We want to allow a lot of things as the user is
    # allowed to note on anything.
    SecRule REQUEST_FILENAME "@contains /index.php/apps/notes/" \
        "id:9508340,\
        phase:1,\
        pass,\
        t:none,\
        nolog,\
        ctl:ruleRemoveByTag=attack-injection-php,\
        ver:'nextcloud-rule-exclusions-plugin/1.0.0'"

Adding the `REQUEST_METHOD' (Value: `PUT' )` here, still throws the FP when saving the Notes settings, because the URI is `"/apps/notes/settings"`.

I have disabled the original rule and added a new one:

    # We want to allow a lot of things as the user is
    # allowed to note on anything.
    SecRule REQUEST_FILENAME "@contains /apps/notes/" \
        "id:9508341,\
        phase:1,\
        pass,\
        t:none,\
        nolog,\
        ctl:ruleRemoveByTag=attack-injection-php,\
        ver:'nextcloud-rule-exclusions-plugin/1.0.0',\
        setvar:'tx.allowed_methods=%{tx.allowed_methods} PUT"

Not only does saving to the settings work, but also the following actions with Notes:

- saving notes
- exotic content like scripts, commands, etc.
- assigning categories
- enable / disable favorites

Would this be a good approach to change that rule? Can you tell me, why the path `/index.php/apps/notes/` apparently works for everything expect saving to settings, but this path `/apps/notes/` works for everything.

Sorry for all those questions and thanks for your help!

Reply:
Sweet! Thats way better. I didn't know one could do that ;)
Are those regex standarized? Can you give me a link for a good start to learn more?

I have implemented that rule and everything is working with notes EXCEPT saving the settings:

    "uri": "/apps/notes/settings"
    "message": "Method is not allowed by policy",
            "details": {
              "match": "Matched \"Operator `Within' with parameter `GET HEAD POST OPTIONS' against variable `REQUEST_METHOD' (Value: `PUT' )",
              "reference": "v0,3",
              "ruleId": "911100",
              "file": "/usr/local/etc/modsecurity/owasp-modsecurity-crs/rules/REQUEST-911-METHOD-ENFORCEMENT.conf",
              "lineNumber": "28",
              "data": "PUT",
              "severity": "2",
              "ver": "OWASP_CRS/4.0.0-rc1",
              "rev": "",
              "tags": [
                "application-multi",
                "language-multi",
                "platform-multi",
                "attack-generic",
                "paranoia-level/1",
                "OWASP_CRS",
                "capec/1000/210/272/220/274",
                "PCI/12.1"
              ],
              "maturity": "0",
              "accuracy": "0"
            }

So i have to add a regex to `SecRule REQUEST_FILENAME "@rx ^(?/index.php)?/apps/notes/` to also catch `settings`?
So we could avoid using `contains`?
Something like you mentioned above: `@rx (?:/index.php)?/apps/notes(?:/settings)?`

Reply:
In general, our regular expressions are Re2 compatible (https://github.com/google/re2/wiki/Syntax).

The rule should have matched. `/app/notes/settings` clearly begins with `/apps/notes/`. But I see that I made an error in the rule. It should be 
```
"@rx ^(?:/index.php)?/apps/notes/"
```
instead of
```
"@rx ^(?/index.php)?/apps/notes/"
```

Reply:
Sounds good. But drop the `(?:/settings)?`. For one, it introduces a double slash. Also, it is superfluous, the matches are identical with and without that piece.

Reply:
Sorry for my late reply @theseion 
And thanks for your help. It works as suggested.
How do we proceed from here?  Do I have to make another pull request with the updated rule or can I change my current one?

Thanks again for your patience. I really appreciate it!

Reply:
Oh, and also we should change the id to: `9508340`
I dont know why it ended up with the id `9508341`

Reply:
Just update this PR.